gc/policy/policy.go

```diff
@@ -0,0 +1,217 @@
+package policy
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/metadata"
+	"github.com/containerd/containerd/plugin"
+)
+
+type config struct {
+	PauseThreshold float64 `toml:"pause_threshold"`
+
+	ScheduleOnDelete  bool `toml:"on_delete"`
+	MutationThreshold int  `toml:"mutation_threshold"`
+
+	ScheduleDelayMs int `toml:"schedule_delay_ms"`
+	StartupDelayMs  int `toml:"startup_delay_ms"`
+}
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		Type: plugin.GCPlugin,
+		ID:   "policy",
+		Requires: []plugin.Type{
+			plugin.MetadataPlugin,
+		},
+		Config: &config{
+			PauseThreshold:    0.02,
+			ScheduleOnDelete:  false,
+			MutationThreshold: 0,
+			ScheduleDelayMs:   0,
+			StartupDelayMs:    100,
+		},
+		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
+			md, err := ic.Get(plugin.MetadataPlugin)
+			if err != nil {
+				return nil, err
+			}
+
+			m := newPolicyManager(md.(*metadata.DB), ic.Config.(*config))
+
+			ic.Meta.Exports = map[string]string{
+				"PauseThreshold":    fmt.Sprint(m.pauseThreshold),
+				"ScheduleOnDelete":  fmt.Sprint(m.onDelete),
+				"MutationThreshold": fmt.Sprint(m.mutationThreshold),
+				"ScheduleDelay":     fmt.Sprint(m.scheduleDelay),
+			}
+
+			go m.run(ic.Context)
+
+			return m, nil
+		},
+	})
+}
+
+type mutationEvent struct {
+	ts    time.Time
+	dirty bool
+}
+
+type collector interface {
+	RegisterMutationCallback(func(bool))
+	GarbageCollect(context.Context) error
+}
+
+type gcPolicyManager struct {
+	c collector
+
+	eventC chan mutationEvent
+
+	pauseThreshold    float64
+	onDelete          bool
+	mutationThreshold int
+	scheduleDelay     time.Duration
+	startupDelay      time.Duration
+}
+
+func newPolicyManager(c collector, cfg *config) *gcPolicyManager {
+	eventC := make(chan mutationEvent)
+
+	m := &gcPolicyManager{
+		c:                 c,
+		eventC:            eventC,
+		pauseThreshold:    cfg.PauseThreshold,
+		onDelete:          cfg.ScheduleOnDelete,
+		mutationThreshold: cfg.MutationThreshold,
+		scheduleDelay:     time.Duration(cfg.ScheduleDelayMs) * time.Millisecond,
+		startupDelay:      time.Duration(cfg.StartupDelayMs) * time.Millisecond,
+	}
+
+	if m.pauseThreshold < 0.0 {
+		m.pauseThreshold = 0.0
+	}
+	if m.pauseThreshold > 0.5 {
+		m.pauseThreshold = 0.5
+	}
+	if m.mutationThreshold < 0 {
+		m.mutationThreshold = 0
+	}
+	if m.scheduleDelay < 0 {
+		m.scheduleDelay = 0
+	}
+	if m.startupDelay < 0 {
+		m.startupDelay = 0
+	}
+
+	c.RegisterMutationCallback(m.mutationCallback)
+
+	return m
+}
+
+func (m *gcPolicyManager) mutationCallback(dirty bool) {
+	e := mutationEvent{
+		ts:    time.Now(),
+		dirty: dirty,
+	}
+	go func() {
+		m.eventC <- e
```

Comment:
~If the order is important, you want to have a queue here. I see that the `mutationEvent` has the time associated with it, so maybe it's not an issue.~

I see that this timestamp is checked later.